src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,295 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_integration.h>
+#include <memory>
+#include <vector>
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+/// \ingroup NumericalAlgorithmsGroup
+/// Numerical integration algorithms
+namespace integration {
+
+namespace detail {
+
+// The GSL functions require the integrand to have this particular function
+// signature. In particular, any extra parameters to the functions must be
+// passed as a void* and re-interpreted appropriately.
+template <typename IntegrandType>
+double integrand(const double x, void* const params) {
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+  const auto function = reinterpret_cast<IntegrandType*>(params);
+  return (*function)(x);
+}
+
+class GslQuadAdaptiveImpl {
+ public:
+  explicit GslQuadAdaptiveImpl(size_t max_intervals) noexcept;
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  void pup(PUP::er& p) noexcept;  // NOLINT(google-runtime-references)
+
+ protected:
+  template <typename IntegrandType>
+  gsl_function* gsl_integrand(IntegrandType&& integrand) const noexcept {
+    gsl_integrand_.function = &detail::integrand<IntegrandType>;
+    gsl_integrand_.params = &integrand;
+    return &gsl_integrand_;
+  }
+
+  struct GslIntegrationWorkspaceDeleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept;
+  };
+
+  size_t max_intervals_ = 0;
+  std::unique_ptr<gsl_integration_workspace, GslIntegrationWorkspaceDeleter>
+      workspace_{};
+
+ private:
+  void initialize() noexcept;
+
+  mutable gsl_function gsl_integrand_{};
+};
+
+void check_status_code(int status_code);
+
+void disable_gsl_error_handling() noexcept;
+
+}  // namespace detail
+
+/// Each type specifies which algortihm from GSL should be used. It should be
```

Comment:
"algorithm"

You may want to look into getting your editor to spell-check your comments.  People became much happier with my PRs after I did that.